### DISCUSSION TITLE
Run custodian across multiple tenants

### DISCUSSION THREAD

**Discussion by rachgupt** [cloud-custodian/cloud-custodian#7388] [General]
I see that subscriptions.yaml when created runs across a tenant and multiple subscriptions on Azure. Can we run it across multiple Azure tenants?

**Comment by kapilt** [comment]
the service principal you run c7n-org with needs access to each subscription, see azure docs on lighthouse and delegated resource management to enable.

**Comment by rachgupt** [comment]
We are using MSI , we could run c7n-org on a tenant which works for all subscriptions on that tenant as we gave access on all subscriptions or management group. What i am seeking to clarify is can we run on different tenant, i.e different AD  itself. Or we need to replicate the environment per tenant?

**Comment by kapilt** [comment]
MSI is a faulty/broken concept imo (it requires post provisioning rbac mgmt, encourages treating resources like pets, and multiples the  iam entities under management), you should use customer assigned identity and then explicitly configure per the azure lighthouse docs for multi-tenant access from the customer assigned identity, per my previous comment, you'll need to read up on multi-tenant access on the azure docs.

**Comment by rachgupt** [comment]
@kapilt Thanks, let me go over the documents.

**Comment by stefangordon** [comment]
So to query the multiple subscription lists we need auth tokens for each tenant.  

With MSI I believe we pull the Tenant ID out of the credentials we are issued, then we request tokens only for that tenant and enumerate subscriptions. 

Potentially this could be setting a tenant ID environment variable before each c7n_org run.  I know it would work fine with Service Principal authentication since you can specify the tenant there manually.  

We'll do some research on it.
